### ISSUE TITLE
Flutter crash while using Camera plugin on iOS

### ISSUE THREAD

**Issue by mehmetf** [flutter/flutter#70577]
With a recent internal roll, customer:money is running into frequent crashes while using the internal camera plugin that seems to step from engine.

```
Exception Type:  EXC_BAD_ACCESS (SIGSEGV)
Exception Subtype: KERN_INVALID_ADDRESS at 0x000000000000000c
VM Region Info: 0xc is not in any region.  Bytes before following region: 4298883060
      REGION TYPE                      START - END             [ VSIZE] PRT/MAX SHRMOD  REGION DETAIL
      UNUSED SPACE AT START
--->  
      __TEXT                 00000001003bc000-0000000103530000 [ 49.5M] r-x/r-x SM=COW  ...os_dev_global
 
Termination Signal: Segmentation fault: 11
Termination Reason: Namespace SIGNAL, Code 0xb
Terminating Process: exc handler [1566]
Triggered by Thread:  0
 
Thread 0 name:  Dispatch queue: com.apple.main-thread
Thread 0 Crashed:
0   Flutter                       	0x0000000107066698 0x106f24000 + 1320600
1   Flutter                       	0x0000000106f55724 0x106f24000 + 202532
2   Flutter                       	0x0000000106f55724 0x106f24000 + 202532
3   Flutter                       	0x0000000107236c34 0x106f24000 + 3222580
4   Flutter                       	0x0000000106f629c4 0x106f24000 + 256452
5   Flutter                       	0x0000000106f650f0 0x106f24000 + 266480
6   CoreFoundation                	0x00000001865393a8 0x18648c000 + 709544
7   CoreFoundation                	0x00000001865390c0 0x18648c000 + 708800
8   CoreFoundation                	0x0000000186538740 0x18648c000 + 706368
9   CoreFoundation                	0x0000000186533538 0x18648c000 + 685368
10  CoreFoundation                	0x0000000186532bc8 0x18648c000 + 682952
11  GraphicsServices              	0x00000001909145cc 0x190911000 + 13772
12  UIKitCore                     	0x000000018a6e5744 0x189c77000 + 10938180
13  ios_dev_global                	0x000000010046abec 0x1003bc000 + 715756
14  libdyld.dylib                 	0x00000001863af384 0x1863ae000 + 4996
```

**Comment by mehmetf** [comment]
@flutter-symbolizer-bot https://github.com/flutter/flutter/issues/70577#issue-743384252 engine#40b8f8227b9a8f4ab78fa9d115cf135aaa885948 ios debug arm64

**Comment by mehmetf** [comment]
Let's try a different format:

```
Thread 0 Crashed:
0x0000000107066698(Flutter + 0x00142698)
0x0000000106f55724(Flutter + 0x00031724)
0x0000000106f55724(Flutter + 0x00031724)
0x0000000107236c34(Flutter + 0x00312C34)
0x0000000106f629c4(Flutter + 0x0003E9C4)
0x0000000106f650f0(Flutter + 0x000410F0)
0x00000001865393a8(CoreFoundation + 0x000AD3A8)
```

**Comment by mehmetf** [comment]
@flutter-symbolizer-bot https://github.com/flutter/flutter/issues/70577#issuecomment-727668624 internal engine#40b8f8227b9a8f4ab78fa9d115cf135aaa885948 ios debug arm64

**Comment by mehmetf** [comment]
OK.. Release builds only.

**Comment by mehmetf** [comment]
@flutter-symbolizer-bot https://github.com/flutter/flutter/issues/70577#issuecomment-727668624 internal engine#40b8f8227b9a8f4ab78fa9d115cf135aaa885948 ios release arm64

**Comment by mehmetf** [comment]
Unfortunately there are too many engine commits in this update:

https://github.com/flutter/engine/compare/c7391d6b12c3...40b8f8227

Perhaps: 0d5f2e9 Rasterizer is initialized with an external view embedder ?

**Comment by mehmetf** [comment]
Another crash; this time in release mode.

```
Thread 0 Crashed:
0x0000000106cff1d8(Flutter + 0x0013F1D8)
0x0000000106bef57c(Flutter + 0x0002F57C)
0x0000000106ecea00(Flutter + 0x0030EA00)
0x0000000106bfc558(Flutter + 0x0003C558)
0x0000000106bfe4cc(Flutter + 0x0003E4CC)
0x00000001865393a8(CoreFoundation + 32)
```

**Comment by mehmetf** [comment]
@flutter-symbolizer-bot https://github.com/flutter/flutter/issues/70577#issuecomment-728104770 internal engine#40b8f8227b9a8f4ab78fa9d115cf135aaa885948 ios release arm64

**Comment by mehmetf** [comment]
@mraleph I am having a bit of trouble morphing the crash report being output by Console to something that symbolizer understands. See the first message as the original format.

**Comment by mehmetf** [comment]
@flutter-symbolizer-bot https://github.com/flutter/flutter/issues/70577#issuecomment-728104770 internal engine#40b8f8227b9a8f4ab78fa9d115cf135aaa885948 ios release arm64

**Comment by mehmetf** [comment]
@flutter-symbolizer-bot https://github.com/flutter/flutter/issues/70577#issuecomment-728104770 internal engine#5f660ce78474b82588258ca0268af5d2684489dd ios release arm64

**Comment by mraleph** [comment]
@mehmetf it's because of case sensitive hex numbers. Fixed.

**Comment by mehmetf** [comment]
@pcsosinski This is going to be a p0 blocker for customer:money soon. Status:

- Their app crashes 9 out of 10 times while grabbing frames from their qr scanner plugin (which is based off of Camera).
- This happens only on iOS.
- It started with a roll that introduced the following engine changes: https://github.com/flutter/engine/compare/c7391d6b12c3...40b8f8227. They are unable to reproduce the crash prior to the roll but can reproduce immediately after the roll.
- They are unable to reproduce the crash in a sample app. They are suspecting it might be related to increased memory usage of their app.
- The symbolized stack is given above: https://github.com/flutter/flutter/issues/70577#issuecomment-728142952

**Comment by chinmaygarde** [comment]
This app seems to using inline texture composition with a platform view in the scene (since the thread that is crashing is the main thread). I am not sure if [Frame 0](https://github.com/flutter/flutter/issues/70577#issuecomment-728142952) of the backtrace is accurate though. We should be able to prepare a reduced test case that has a webview as well as a camera view in the same scene to reproduce conditions similar to the ones observed in the app.

Assigning to @iskakaushik for investigation.

**Comment by iskakaushik** [comment]
On it, crafting a repro case.

**Comment by iskakaushik** [comment]
@mehmetf apparently my version of xcode doesn't support iOS 14.2. I'm updating it and it will be a little bit before I can repro this issue. I will aim to have an update by EOD.

**Comment by iskakaushik** [comment]
I was able to reproduce this using the following example: https://gist.github.com/iskakaushik/fa64ed87732e2b6ffeddf9e4cf0b7104. I have a fix for the issue for this at https://github.com/flutter/engine/pull/22557.

There were several other issues that I encountered during this:
1. Camera plugin registers external textures on a non-platform thread this causes failures in weak-ptr thread affinity checking. See: https://github.com/flutter/flutter/issues/69933. Camera plugin straight-up doesn't work in debug engine builds.

2. We call submit_frame_ with an empty canvas in gpu_surface_metal.mm. I am not sure why this gets called. I'm worried about any surfaces that are non un-refd. Will file an issue for this.

**Comment by cyanglaz** [comment]
The root cause is that when redrawing the last layer tree, composition_order.clear() is not called. It is a regression caused by https://github.com/flutter/engine/pull/22372 where I moved the composition_order.clear() out of SubmitFrame.

We need to make sure EndFrame is called in all the frames, in either regular frame or redrawing frame.

**Comment by wanjm** [comment]
will this be pickup in 1.22, I come across this too;

**Comment by wanjm** [comment]
when will it 1.22 cherry pick ready, I also come accross this when using webview_flutter;

**Comment by wanjm** [comment]
we plan to publish our app this weekend, but is blocked by this bug,  @cyanglaz when will the cherry pickup ready, or any other work around for this. 
1.22.1 1.22.2 also has other problem for us.  #66968 
is there any other branch we can use;

**Comment by KiritoBeater** [comment]
maybe same problem,  let me try symbolizer-bot
```
14  Flutter                       	0x0000000103726a58 0x103710000 + 92760
15  Flutter                       	0x0000000103726a58 0x103710000 + 92760
16  Flutter                       	0x0000000103740718 0x103710000 + 198424
17  Flutter                       	0x0000000103a145cc 0x103710000 + 3163596
18  Flutter                       	0x000000010374b43c 0x103710000 + 242748
19  Flutter                       	0x000000010374d3b0 0x103710000 + 250800
```
@flutter-symbolizer-bot force ios arm64 engine#2c956a31c0a3d350827aee6c56bb63337c5b4e6e this

**Comment by KiritoBeater** [comment]
> [flutter-symbolizer-bot](https://github.com/flutter-symbolizer-bot/flutter-symbolizer-bot/blob/main/README.md#commands) 👉 I only answer to commands of public members of Flutter GitHub org!

I didn't see this. So I can't use it 😭

**Comment by mraleph** [comment]
@flutter-symbolizer-bot force ios arm64 engine#2c956a31c0a3d350827aee6c56bb63337c5b4e6e https://github.com/flutter/flutter/issues/70577#issuecomment-733457351

**Comment by mraleph** [comment]
@flutter-symbolizer-bot force ios arm64 engine#2c956a31c0a3d350827aee6c56bb63337c5b4e6e https://github.com/flutter/flutter/issues/70577#issuecomment-733457351

**Comment by KiritoBeater** [comment]
> @flutter-symbolizer-bot force ios arm64 engine#2c956a31c0a3d350827aee6c56bb63337c5b4e6e [#70577 (comment)](https://github.com/flutter/flutter/issues/70577#issuecomment-733457351)

Thank you so much. This is the same problem.

**Comment by pcsosinski** [comment]
FYI 74d40c1 will require a manual merge into flutter-1.22-candidate.12 cc @christopherfujino @cyanglaz
